docs/design-docs/benchmarks.md

```diff
@@ -0,0 +1,274 @@
+---
+id: benchmarks
+---
+
+# Benchmarks Design
+
+## Goals
+
+The design of benchmarks has the following goals to achieve:
+
+**High Code Quality**
+* Extract the common code into the base class, reduce the workload to maintain all benchmarks.
+* Provide a unified entrance to launch benchmarks.
+
+**Good Extensibility**
+* Avoid modifying existing code when add new benchmarks by using registration mechanism.
+* Support benchmark-specific configuration, further support models with (same structure and different configs).
+
+**Good Usability**
+* Unify the output format for all the micro-benchmarks and E2E-benchmarks, contains return_code, metrics, raw-output, etc.
+* Can specify the run times for user to check the variance of one benchmark.
+
+## High-level Technical Design
+
+![Structure of Benchmark Package](../assets/benchmark-structure.png)
+*Figure 1 Structure of Benchmarks package*
+
+The structure of Benchmarks package can be divided into layers from the bottom up:
+1. Abstract base classes for all kind of benchmarks, including Benchmark, ModelBenchmark, Microbenchmark and DockerBenchmark.
+   1. Benchmark is the base class for all benchmarks. It defines common interfaces such as run(), _preprocess(), _postprocess(), _benchmark(), add_parser_arguments() and so on.
+   2. ModelBenchmark is the base class for all E2E models. It defines the abstract interfaces that need to be implemented by the subclasses using different frameworks, such as PytorchBase, TFBase and ONNXBase. Each subclass will realize part of the abstract interfaces that is common for models, such as _init_distributed_setting(), _init_dataloader(), _create_optimizer().
+   3. Microbenchmark is the base class for all micro benchmarks. It defines the abstract interfaces that need to be implemented by the subclasses, such as _process_raw_result(), _process_numeric_result().
+   4. DockerBenchmark is the base class for real workloads based on docker. It also defines the abstract interfaces that need to be implemented by the subclasses.
+2. Derived classes for all implemented benchmarks, realized all the abstract interfaces. The benchmarks will be registered into Benchmark Registry.
+3. BenchmarkRegistry: provides a way of benchmark registration, maintains all the registered benchmarks and supports benchmark launching by BenchmarkContext.
```

Comment:
`BenchmarkRegistry` provides a way of benchmark registration, maintains all the registered benchmarks, and supports benchmark launching by `BenchmarkContext`.

Reply:
Done